include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1270 @@
+              const std::vector<ScalarType> &independent_variables,
+              Vector<ScalarType>            &values);
+
+      /**
+       * Computes the gradient of the vector field.
+       *
+       * @param[in] active_tape_index The index of the tape on which the dependent
+       *            function is recorded.
+       * @param[in] n_dependent_variables The number of dependent variables.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] jacobian The scalar values of the dependent function's jacobian.
+       */
+      static void
+      jacobian (const unsigned int            &active_tape_index,
+                const unsigned int            &n_dependent_variables,
+                const std::vector<ScalarType> &independent_variables,
+                FullMatrix<ScalarType>        &jacobian);
+
+      //@}
+
+    };
+
+
+
+    /**
+     * A prototype driver class for tapeless auto-differentiable numbers.
+     *
+     * It is intended that this class be specialized for the valid
+     * combinations of auto-differentiable numbers and output scalar
+     * number types.
+     *
+     * @tparam ADNumberType A type corresponding to a supported
+     *         auto-differentiable number.
+     * @tparam ScalarType A real or complex floating point number.
+     * @tparam T An arbitrary type resulting from the application of
+     *         the SFINAE idiom to selectively specialize this class.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template<typename ADNumberType, typename ScalarType, typename T = void>
+    struct TapelessDrivers
+    {
+      // This dummy class definition safely supports compilation
+      // against taped numbers or tapeless number types that have
+      // not yet been implemented.
+
+      /**
+       * @name Configuration
+       */
+      //@{
+
+      /**
+       * In the event that the tapeless mode requires <i>a priori</i> knowledge
+       * of how many directional derivatives might need to be computed, this function
+       * informs the auto-differentiable library of what this number is.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       *            that will be used in the entire duration of the
+       *            simulation.
+       *
+       * @warning For Adol-C tapeless numbers, the value given to @p n_independent_variables
+       *          should be the <b>maximum</b> number of independent variables that will be
+       *          used in the entire duration of the simulation. This is important in the
+       *          context of, for example, hp-FEM and for multiple constitutive models with
+       *          a different number of fields from which a linearization must be computed.
+       */
+      static void
+      initialize (const unsigned int &n_independent_variables);
```

Comment:
New function

Reply:
same question here about how these `static` member functions are supposed to work: an `initialize()` function must initialize *something*, but because the function is `static` it can't be an object of the surrounding type, and because the argument is `const`, that's not it either. 

Also: no reference necessary. Same below.

Reply:
I suppose I could rename this to `initialize_global_environment` or something like that. Would this be better?

Reply:
Yes, I think that would make the intent clearer!